### PR TITLE
:bug: Skip HelmChartProxy reconciliation for paused clusters

### LINKED ISSUE #188: Duplicate HelmReleaseProxy resources after move - potential race
Can you help me understand why this reconciles on `ClusterUnpaused` OR `ClusterControlPlaneInitialized` and not on both of those (AND)? I've seen what looks like a race on move where I find multiple `HelmReleaseProxies` for the same `HelmChartProxy` and cluster - labels are set up correctly and created at same time (albeit at 1s fidelity from `CreationTimestamp` so can't tell which is created first).

~I think that changing this logic to when a cluster is not paused AND the control plane has been initialized would solve this problem.~ I see that wouldn't work because the ClusterUnpaused explicitly requires the cluster to have previously been in unpaused state before to continue on update, which wouldn't be the case just following cluster control plane initialized.

From the move logs you can see the order that causes this issue:

```
Creating Cluster="self-hosted-18gxrd" Namespace="self-hosted-wokpvl"
...
Creating HelmChartProxy="cluster-autoscaler-self-hosted-18gxrd" Namespace="self-hosted-wokpvl"
...
Creating HelmReleaseProxy="cluster-autoscaler-self-hosted-18gxrd-7q6p4" Namespace="self-hosted-wokpvl"
```

By the time the `HelmReleaseProxy` has been moved, a new `HelmReleaseProxy` has already been created it seems.

### PR DESCRIPTION
**Pull request by jimmidyson**
This fixes a bug leading to duplicate `HelmReleaseProxy` resources
following a `clusterctl move` that can happen due to a race between
reconciling the `HelmChartProxy` and moving and unpausing the `Cluster`.

Fixes #188.

### CODE REVIEW

**Comment by jimmidyson** [comment]
Tested locally and this fixed my issue btw.

**Review by mboersma — COMMENTED** [review]
Thanks for this! It does seem like a fix.

Would you be up for writing a unit test that captures this behavior in `helmchartproxy_controller_test.go`? It should be mostly copy-and-paste to add a new Cluster definition with `Spec.Paused` == true, then a new test case that ensures the HCP is Ready (and that it isn't ready without your change).

**Comment by jimmidyson** [comment]
@mboersma On it! Thanks for your review!

File: controllers/helmchartproxy/helmchartproxy_controller.go [hunk 1]
```diff
@@ -218,6 +218,12 @@ func (r *HelmChartProxyReconciler) reconcileNormal(ctx context.Context, helmChar
 			continue
 		}
 
+		// Don't reconcile if the Cluster is paused.
+		if cluster.Spec.Paused {
```
Comment (Jont828) [inline controllers/helmchartproxy/helmchartproxy_controller.go:222]:
```suggestion
		if annotations.IsPaused(cluster, cluster) {
```
Seems like the convention in CAPI is to use this annotation getter.

File: controllers/helmchartproxy/helmchartproxy_controller.go [hunk 2]
```diff
@@ -218,6 +218,12 @@ func (r *HelmChartProxyReconciler) reconcileNormal(ctx context.Context, helmChar
...
+			log.V(2).Info("Cluster is paused, skipping reconcile", "cluster", cluster.Name)
```
Comment (Jont828) [inline controllers/helmchartproxy/helmchartproxy_controller.go:223]:
```suggestion
			log.V(2).Info("Reconciliation is paused for this Cluster", "cluster", cluster.Name)
```

**Review by Jont828 — COMMENTED** [review]
Small nits, but otherwise a pretty straightforward change.

File: controllers/helmchartproxy/helmchartproxy_controller.go (hunk 1 above)
Comment (jimmidyson) [reply controllers/helmchartproxy/helmchartproxy_controller.go:222]: Ah good tip! Just a small adjustment to check the paused annotation on the `helmChartProxy` and the `cluster.Spec.Paused` via the call you suggest. I'm also going to move this into `reconcileForCluster` which I think makes more sense.

File: controllers/helmchartproxy/helmchartproxy_controller.go (hunk 2 above)
Comment (jimmidyson) [reply controllers/helmchartproxy/helmchartproxy_controller.go:223]: Done, thanks!

**Comment by jimmidyson** [comment]
Added test cases, tested the HRP gets created without this change, and with this change does not. Also added test case to ensure when a cluster is unpaused the HRP gets created.

**Comment by Jont828** [comment]
Please squash your commits, otherwise looks good!

**Comment by jimmidyson** [comment]
@Jont828 @mboersma Squashed! Please lgtm approve when you get a moment 🙏 

And if you could cut a release once this is merged that would be most appreciated, it's blocking some of our e2e tests.

**Review by mboersma — APPROVED** [review]
/lgtm

**Review by jackfrancis — APPROVED** [review]
/lgtm
/approve

**Comment by mboersma** [comment]
@jimmidyson unfortunately @Jont828 is away for several days, and the other maintainers aren't familiar with the release process. 

We have a bus factor problem here. I'll create an issue for the lack of release documentation. (If you or someone you know would like to be a reviewer or maintainer on CAAPH, I think we would entertain that idea too.)

**Comment by jimmidyson** [comment]
@mboersma Thanks for the update. I'd happily help out as a maintainer if you'll have me 😅

**Comment by mboersma** [comment]
Thanks @jimmidyson! Let's wait until later next week when the current maintainers are all available and I'll propose it.

**Comment by Jont828** [comment]
@jimmidyson Hey, I'm back from vacation. Super glad to hear that you're interested in getting involved! I'll reach out on Slack if you'd like to chat more.

**Pull request merged**